File: app/src/main/java/com/example/android/newsapp/refactor/repository/remote/NewsRemoteRepository.java
package com.example.android.newsapp.refactor.repository.remote;

import com.example.android.newsapp.refactor.model.News;
import com.example.android.newsapp.refactor.repository.NewsGateway;
import com.example.android.newsapp.refactor.repository.remote.api.NewsService;
import com.example.android.newsapp.refactor.usecase.DataCallback;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

//TODO Implements remote layer
public class NewsRemoteRepository implements NewsGateway {

    private final NewsService newsService;

    public NewsRemoteRepository(NewsService newsService) {
        this.newsService = newsService;
    }

    @Override
    public void getNews(final DataCallback dataCallback) {
        newsService.getNews().enqueue(new Callback<News>() {
            @Override
            public void onResponse(Call<News> call, Response<News> response) {
                dataCallback.onSuccess(response.body().getResponse().getResults());
            }

            @Override
            public void onFailure(Call<News> call, Throwable t) {

            }
        });
    }
}

File: app/src/main/kotlin/com/example/android/newsapp/repository/remote/NewsRemoteRepository.kt
package com.example.android.newsapp.repository.remote

import com.example.android.newsapp.model.News
import com.example.android.newsapp.model.Result
import com.example.android.newsapp.repository.NewsGateway
import com.example.android.newsapp.repository.api.NewsService
import com.example.android.newsapp.usecase.base.DataCallback
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class NewsRemoteRepository(private val service: NewsService) : NewsGateway {

    override fun getNews(dataCallback: DataCallback<List<Result>>) {
        service.getNews().enqueue(object : Callback<News> {
            override fun onResponse(call: Call<News>, response: Response<News>) {
                dataCallback.onSuccess(response.body()?.response?.results!!)
            }

            override fun onFailure(call: Call<News>, throwable: Throwable) {
            }
        })
    }
}
